### DISCUSSION TITLE
How to write/read Flash?

### DISCUSSION THREAD

**Discussion by gege0500** [polarfire-soc/polarfire-soc-documentation#290] [Applications and demos]
polarfire-soc-spi-flash-example:
https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mss-spi/mpfs-spi-flash#polarfire-soc-spi-flash-example
It is mentioned above:"This project can not be directly used on the Icicle kit since the flash memory device is not accessible via SPI."
The code was programmed and debugged. When debugging, the process is always trapped here:(mss_spi.c #531) while(transfer_idx<transfer_size).
How to write/read Flash?

**Comment by TommyMurphyTM1234** [comment]
Well, you obviously need a SPI flash connected to the pins that the software uses. The Icicle board doesn't provide that.

**Reply by gege0500** [reply]
The Icicle board provides a 1Gb flash (MT25QL01GBBB8ESF-0SIT). How to use it? Why can't the spi flash demo drive this flash？

**Comment by hughbreslin** [comment]
Hey @gege0500 there are two ways of using spi with PolarFire SoC:

1. the way @TommyMurphyTM1234 mentioned above is the MSS peripheral SPI (this could be with the MSS SPI controller or a fabric CoreSPI) and connect a SPI part to I/Os 
2. System controller spi, this is a separate spi interface which is used by the FPGA’s system controller for in application programming (IAP), field updates or data storage. This is what the MT25QL01GBBB8ESF-0SIT is connected to. You cannot access this using the SPI driver, you need to use system services. Have a look at this [document](https://www.microsemi.com/document-portal/doc_download/1245815-polarfire-fpga-and-polarfire-soc-fpga-system-services-user-guide). There is a Linux and bare metal example for system services but one thing to note is that on ES parts (I.e the part on an icicle kit) you cannot write to the system controller SPI from the MSS but you can program it from libero. Therefore on ES the system controller SPI is basically ROM for the MSS.

**Reply by gege0500** [reply]
Thank you for your reply. I see that there is a section about SPI flash programming in the Libero manual (https://onlinedocs.microchip.com/oxy/GUID-AFCB5DCC-964F-4BE7-AA46-C756FA87ED7B-en-US-4/GUID-44FADDD5-B7E5-45C1-98B0-A925B8734666.html). To store data to SPI flash, a hex/ bin file is needed uploaded to the Add Data Storage Client. But I don't know how to generate hex/bin files that can store data to SPI flash? Can you provide an example please? Looking forward to your reply.

**Reply by hughbreslin** [reply]
I've done this with Python before, see the sample below:

```python
    input_file = open("input_data.text", mode="rb")
    data = input_file.read()
    input_file.close()
    output_file = open("./spi-client.bin", "wb+")
    output_file.write(data)
    output_file.close()
```

There is nothing special about the hex/bin format - the main thing is what data you want to store. For the example above it converts a text file to bytes and writes it to a .bin file where I can then unpack it on the system. You could do this with any file format. If you wanted to store just plan numerical values you could just convert them directly to bytes instead of storing them in a file and write that data to a .bin.

**Reply by gege0500** [reply]
Thank you for your reply. Now I understand more clearly. But I still don't understand what the "input_data.text" you mentioned above looks like? For example, I want to store 0x12 in 00002000H address. How should I express it in text? Can you provide a demo please? Looking forward to your reply.

**Reply by hughbreslin** [reply]
Ahh ok - my example stores the file (and obviously the file contents but it doesn't extract any data from the file). A better example would be:

```python
    my_value = 0x12
    data = my_value.to_bytes(byteorder="little", length=4)
    output_file = open("./spi-client.bin", "wb+")
    output_file.write(data)
    output_file.close()
```

`length=4` specifies the number of bytes.

This will generate a client with no addressing. when you add the client to the Libero project you can specify the start address, e.g:

![image](https://user-images.githubusercontent.com/44862881/220946988-2cc4ecec-9d1e-48f5-8162-9e45527e4862.png)
